### PR TITLE
Fix timeouts and split node-freddy

### PR DESCRIPTION
**Pull request by indrekj**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,59 +22,30 @@ freddy.deliver(destination, message)
...
-    * The parsed response
-
-    * The `MessageHandler`(described further down)
-
-* Synchronous deliver expecting response
+  * If `timeout` seconds pass without a response then `Freddy::ErrorResponse` error is raised with #response
 ```ruby
-  response = freddy.deliver_with_response(destination, message, timeout_seconds = 3)
+rescue Freddy::ErrorResponse => e
+  e.response # => { error: 'Timed out waiting for response' }
```
Comment (urmastalimaa) [inline README.md]: Why is the error under the `response` method? Maybe put it under `error`?

Comment (urmastalimaa) [inline README.md]: Or possibly `reason`?

Comment (indrekj) [inline README.md]: It can be whatever you want. e.g
`msg_handler.error(id: 3)`
then
`e.response # => { id: 3 }`

Comment (indrekj) [inline README.md]: if you mean just the method name then `exception.error` seemed a bit odd. especially if most of our messages are `exception.error # => {error: 'something'}`

Comment (urmastalimaa) [inline README.md]: got it

File: lib/freddy/message_handlers.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
+class Freddy
+  module MessageHandlers
+    def self.for_type(type)
+      type == 'request' ? RequestHandler : StandardMessageHandler
+    end
+
+    class BaseMessageHandler
```
Comment (urmastalimaa) [inline lib/freddy/message_handlers.rb]: I don't really see the point of this base class.

Comment (urmastalimaa) [inline lib/freddy/message_handlers.rb]: Is it for documenting the API?

Comment (indrekj) [inline lib/freddy/message_handlers.rb]: pretty much. the old code also had this class. I can remove it. doesn't really give anything beside the initializer

Comment (urmastalimaa) [inline lib/freddy/message_handlers.rb]: yeah I think it can be removed

File: lib/freddy/request.rb [hunk 3]
```diff
@@ -22,30 +21,50 @@ def initialize(channel, logger)
       @producer, @consumer = Producer.new(channel, logger), Consumer.new(channel, logger)
       @listening_for_responses = false
       @request_map = Hamster.mutable_hash
+      @timeout_clearer = RequestTimeoutClearer.new @request_map, @logger
+
+      @producer.on_return do |return_info, properties, content|
+        if return_info[:reply_code] == NO_ROUTE
+          @timeout_clearer.force_timeout(properties[:correlation_id])
```
Comment (urmastalimaa) [inline lib/freddy/request.rb]: It this actually a timeout? When exactly is NO_ROUTE returned?

Comment (indrekj) [inline lib/freddy/request.rb]: it is not an actual timeout. this happens when the queue doesn't exist at all. previously freddy waited `timeout` seconds and then returned the timeout error. I made it faster by timing out immediately. In reality, it is not a real timeout error, but I'm not sure what's the better way to handle it.

Comment (urmastalimaa) [inline lib/freddy/request.rb]: Maybe the timeout can be cleared and the we can invoke the callback with the correct error?

**Comment by urmastalimaa** [comment]
Update gemspec as well

**Comment by indrekj** [comment]
Updated

**Pull request merged**